From 5e7bd5815e56326d9308daa12b31efd06a3d2ddd Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 15 Jun 2023 09:39:57 +0800
Subject: [PATCH 49/53] [Huawei] mm/sharepool: Update sp_init_group_master
 process

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

We need only initilize the sp_group_master for a task in sp_alloc and
k2u for the SPG_ID_DEFAULT. Because the sp_group_master would have
been initilized in the add group process.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 39 +++++++++++++++++++++++----------------
 1 file changed, 23 insertions(+), 16 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 06eabcd8f0b4..defb5bee6445 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -574,6 +574,24 @@ static int sp_init_group_master_locked(struct task_struct *tsk, struct mm_struct
 	return ret;
 }
 
+static int sp_init_group_master(struct task_struct *tsk, struct mm_struct *mm)
+{
+	int ret;
+	down_read(&sp_global_sem);
+	/* The sp_group_master would never change once set */
+	if (mm->sp_group_master) {
+		up_read(&sp_global_sem);
+		return 0;
+	}
+	up_read(&sp_global_sem);
+
+	down_write(&sp_global_sem);
+	ret = sp_init_group_master_locked(tsk, mm);
+	up_write(&sp_global_sem);
+
+	return ret;
+}
+
 static struct sp_group *sp_get_local_group(struct task_struct *tsk, struct mm_struct *mm)
 {
 	int ret;
@@ -1273,14 +1291,9 @@ static int mm_add_group_init(pid_t tgid, struct mm_struct **pmm)
 		goto out_put_task;
 	}
 
-	if (!mm->sp_group_master) {
-		/* We don't free mm->sp_group_master on failure */
-		down_write(&sp_global_sem);
-		ret = sp_init_group_master_locked(tsk, mm);
-		up_write(&sp_global_sem);
-		if (ret)
-			goto out_put_mm;
-	}
+	ret = sp_init_group_master(tsk, mm);
+	if (ret)
+		goto out_put_mm;
 
 	if (mm->sp_group_master && mm->sp_group_master->tgid != tgid) {
 		pr_err("add: task(%d) is a vfork child of the original task(%d)\n",
@@ -1989,14 +2002,11 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 		sp_flags |= SP_HUGEPAGE;
 
 	if (spg_id == SPG_ID_DEFAULT) {
-		int ret;
 		/*
 		 * We should first init the group_master in pass through scene and
 		 * don't free it until we release the mm_struct.
 		 */
-		down_write(&sp_global_sem);
-		ret = sp_init_group_master_locked(current, current->mm);
-		up_write(&sp_global_sem);
+		int ret = sp_init_group_master(current, current->mm);
 		if (ret) {
 			pr_err("sp_alloc init local mapping failed %d\n", ret);
 			return ret;
@@ -2385,10 +2395,7 @@ static int sp_k2u_prepare(unsigned long kva, unsigned long size,
 	}
 
 	if (spg_id == SPG_ID_DEFAULT) {
-		int ret;
-		down_write(&sp_global_sem);
-		ret = sp_init_group_master_locked(current, current->mm);
-		up_write(&sp_global_sem);
+		int ret = sp_init_group_master(current, current->mm);
 		if (ret) {
 			pr_err("k2u_task init local mapping failed %d\n", ret);
 			return ret;
-- 
2.17.1

